session/session.go

```diff
@@ -2843,30 +2844,77 @@ func logStmt(execStmt *executor.ExecStmt, s *session) {
 	}
 }
 
-func logGeneralQuery(execStmt *executor.ExecStmt, s *session, isPrepared bool) {
+func genFnGetQuery(execStmt *executor.ExecStmt, s *session, isPrepared bool) func(buf *strings.Builder) string {
 	vars := s.GetSessionVars()
-	if variable.ProcessGeneralLog.Load() && !vars.InRestrictedSQL {
-		var query string
+	return func(buf *strings.Builder) string {
 		if isPrepared {
-			query = execStmt.OriginText()
+			buf.WriteString(execStmt.OriginText())
 		} else {
-			query = execStmt.GetTextToLog()
+			buf.WriteString(execStmt.GetTextToLog())
 		}
 
-		query = executor.QueryReplacer.Replace(query)
+		bufString := buf.String()
+		queryMutable := *(*[]byte)(unsafe.Pointer(&bufString))
+		for i, b := range queryMutable {
+			if b == '\r' || b == '\n' || b == '\t' {
+				queryMutable[i] = ' '
+			}
+		}
 		if !vars.EnableRedactLog {
-			query += vars.PreparedParams.String()
+			buf.WriteString(vars.PreparedParams.String())
 		}
-		logutil.BgLogger().Info("GENERAL_LOG",
-			zap.Uint64("conn", vars.ConnectionID),
-			zap.Stringer("user", vars.User),
-			zap.Int64("schemaVersion", s.GetInfoSchema().SchemaMetaVersion()),
-			zap.Uint64("txnStartTS", vars.TxnCtx.StartTS),
-			zap.Uint64("forUpdateTS", vars.TxnCtx.GetForUpdateTS()),
-			zap.Bool("isReadConsistency", vars.IsIsolation(ast.ReadCommitted)),
-			zap.String("current_db", vars.CurrentDB),
-			zap.String("txn_mode", vars.GetReadableTxnMode()),
-			zap.String("sql", query))
+		return buf.String()
+	}
+}
+
+func genGetUser(user *auth.UserIdentity) func() string {
+	return func() string {
+		if user == nil {
+			return ""
+		}
+		var buf strings.Builder
+		buf.Grow(64)
+		buf.WriteString(user.Username)
+		buf.WriteByte('@')
+		buf.WriteString(user.Hostname)
+		return buf.String()
+	}
+}
+
+func logGeneralQuery(execStmt *executor.ExecStmt, s *session, isPrepared bool) {
+	vars := s.GetSessionVars()
+	if variable.ProcessGeneralLog.Load() && !vars.InRestrictedSQL {
+		fnGetQuery := genFnGetQuery(execStmt, s, isPrepared)
+		fnGetUser := genGetUser(vars.User)
+		fnGetSMV := func() int64 {
+			return s.GetInfoSchema().SchemaMetaVersion()
+		}
+
+		logEntry := logutil.GeneralLogEntry{
+			ConnID:                 vars.ConnectionID,
+			FnGetUser:              fnGetUser,
+			FnGetSchemaMetaVersion: fnGetSMV,
+			TxnStartTS:             vars.TxnCtx.StartTS,
+			TxnForUpdateTS:         vars.TxnCtx.GetForUpdateTS(),
+			IsReadConsistency:      vars.IsIsolation(ast.ReadCommitted),
+			CurrentDB:              vars.CurrentDB,
+			TxnMode:                vars.GetReadableTxnMode(),
+			FnGetQuery:             fnGetQuery,
+		}
+
+		logutil.PutGeneralLogOrDrop(&logEntry)
```

Comment:
Maybe we can have a sync pool for this to avoid frequent allocation

Reply:
done.